Kernel/FileSystem/FATFS/FileSystem.cpp

```diff
@@ -10,6 +10,77 @@
 
 namespace Kernel {
 
+DOSBIOSParameterBlockVersion DOSBIOSParameterBlock::get_bpb_version() const
+{
+    // A DOS 7 EBPB should _never_ contain the values 0x28 or 0x29 at
+    // the offset associated with `dos4_block->signature`
+    // (aka `dos7_block->sectors_per_fat_32bit`) due to the maximum number of
+    // clusters ensuring the number of sectors per fat will not exceed 0x200000.
+    // As a result, it should be safe to determine BPB version through the
+    // signature fields by checking the DOS 4 signature offset prior to the DOS 7 one.
```

Comment:
DOS 7 EBPB having 0x28 or 0x29 at `dos4_block->signature` is one of those "should never happen" occurrences, we should at least check and log when it happens, or return an `ErrorOr<Version>` so that calling code can fail to mount the filesystem if it violates our assumptions. Hardware and legacy software have a tendency to not care about your assumptions or specs.

Reply:
I don't think I'm quite grasping what the proposed criteria for returning an error wold be here.

In the current implementation, the `else if (dos7_block->signature == 0x28 || dos7_block->signature == 0x29)` conditional is only checked _when_ `dos4_block->signature` does **not** contain 0x28/0x29, as those values are checked first by the earlier `if` statements.

One option could be to check that _either_ `dos4_block->signature` _OR_ ` dos7_block->signature` contain the expected value (but not both). However, unlike it being invalid for the DOS 7 EBPB to contain `>0x20` at `dos4_block->signature`, there isn't a guarantee that for a DOS 4 EBPB `dos7_block->signature` will have any particular value. The address of `dos7_block->signature` (`0x42` in the sector) is in unallocated/unused space for a DOS 4 EBPB, so it could contain any arbitrary value.

Reply:
All that said, at a minimum, we can definitely log a warning if both dos7 and dos4 blocks contain valid signatures. I don't think that should prevent _mounting_ the device, however.

Reply:
Warning added: 

https://github.com/tajmorton/serenity/blob/8c42e9a7134a27c5103f1fab2f1949dcdb774134/Kernel/FileSystem/FATFS/FileSystem.cpp#L15-L30

Reply:
> I don't think I'm quite grasping what the proposed criteria for returning an error wold be here.

I don't think I fully understood it when I made the comment either 😅 . But as a general rule, if we are making assumptions about the hardware, we should document when there's a violation with a warning log message, or with bailing out of whatever hardware transaction we were trying to do or initialize.